File: tools/calibrate/testdata/corpus/csharp/Result.cs
namespace Example.Calibration;

public readonly record struct Result<T>(T? Value, string? Error)
{
    public bool IsSuccess => Error is null;

    public static Result<T> Success(T value) => new(value, null);

    public static Result<T> Failure(string message) =>
        new(default, string.IsNullOrWhiteSpace(message)
            ? throw new ArgumentException("An error message is required.", nameof(message))
            : message);

    public Result<TResult> Map<TResult>(Func<T, TResult> transform)
    {
        ArgumentNullException.ThrowIfNull(transform);
        return IsSuccess
            ? Result<TResult>.Success(transform(Value!))
            : Result<TResult>.Failure(Error!);
    }

    public T Unwrap() => IsSuccess
        ? Value!
        : throw new InvalidOperationException($"Cannot unwrap failed result: {Error}");
}

File: tools/calibrate/testdata/corpus/typescript/result.ts
export type Result<T, E> =
  | { readonly ok: true; readonly value: T }
  | { readonly ok: false; readonly error: E };

export const success = <T>(value: T): Result<T, never> => ({ ok: true, value });
export const failure = <E>(error: E): Result<never, E> => ({ ok: false, error });

export function map<T, U, E>(result: Result<T, E>, transform: (value: T) => U): Result<U, E> {
  return result.ok ? success(transform(result.value)) : result;
}

export class ValidationError extends Error {
  constructor(
    readonly field: string,
    message: string,
  ) {
    super(`${field}: ${message}`);
    this.name = "ValidationError";
  }
}

export function parsePort(value: unknown): Result<number, ValidationError> {
  const port = typeof value === "string" ? Number.parseInt(value, 10) : Number(value);
  if (!Number.isInteger(port) || port < 1 || port > 65_535) {
    return failure(new ValidationError("port", `expected 1..65535, received ${String(value)}`));
  }
  return success(port);
}
